### DISCUSSION TITLE
Triaging and specifying major feature for Jamulus 4

### DISCUSSION THREAD

**Discussion by ann0see** [jamulussoftware/jamulus#3382] [Ideas]
In https://github.com/orgs/jamulussoftware/discussions/3341 we discussed potential larger features for Jamulus 4.
This discussion serves as triaging the most important and actionable feature we can implement for Jamulus 4. 

Step 1: Decide on primary feature to focus on
Step 2: Plan how to get this feature going. 

The best outcome would be to have an actionable plan as issue on what to work on exactly in form of a checklist.

We have 4 larger feature requests:

1. Provide uncompressed/higher quality audio.
  ~2. Add multi channel support~
3. Implement Access control functionality
4. Introduce plugin compatibility in Jamulus

**Comment by ann0see** [comment]
Thread for improved audio quality

**Reply by ann0see** [reply]
There have been multiple threads on improved audio quality. 

I think that we should rather focus on improving audio quality issues due to jitter. This is hard to achieve. 

I don't know if a higher sampling rate or less compression is really that valuable? Is this really a major improvement?

**Reply by pljones** [reply]
We'd need to do some detailed investigation of what "audio quality issues due to jitter" means, first, I think.

What's happening?  Is it a problem that's easy to reproduce so one can see where in the code it could be happening and whether a change has made an improvement?

**Reply by ann0see** [reply]
I think it was just a general investigation here: https://github.com/orgs/jamulussoftware/discussions/2912

**Reply by pljones** [reply]
Okay, tagging @hselasky for this, then.  Also @softins as I think you looked at the audio packet re-ordering?

Here's a rough outline of what I think's involved to maintain backwards compatibility:

`-s -p <port number>` gives "base port" to the server, as now.  Add `--multiport <max ports>` to listen on `<port number>` through `<port number> + <max ports> - 1` and turn on the new feature.

`-p <port number>` on the client works as it does now (as `-p` is the port or ports the client or server listen on).

Server registration with a directory is unchanged - base port only.

The client initial connection with the server is unchanged, using the base port only.

However, a new server response message to the client is then needed to indicate "you can also use the extra ports" (including the low and high port numbers in the message).

A client can choose how to send packets using these additional ports
- stick to current approach of using only the base port
- sending the same packet to every port in the range
- sending a new packet to each port in round robin or some scattering approach
- sending the same packet to several ports before moving on to the next packet sent on different ports
- or some other approach entirely of the client's choosing (subject to sane packet re-ordering at the server being possible!).

We'll probably want to start simple - either "old-style" single port, "single port, round robin" or "flood" (sending the same packet to all ports).  Once the server side has settled into that, tuning the scattering algorithm can be looked into further.

A server that has indicated additional ports are open handles any and all combinations of packets arriving, reorders by sequence number, de-duplicates any already received (dropping silently the duplicates).  Other than having sixteen ports rather than one per client channel to queue incoming audio packets from, this isn't really a huge change, I think.

**Reply by softins** [reply]
OK, I hadn't seen the discussion from HPS before, about using multiple ports. Just read it, but need to read it again to fully understand the principle.

I did the initial proof of concept for packet ordering, in order to convince Volker that it was needed and a good idea. He then implemented the principle in a different (and better) way, as he knew the operation of the code much better than I did. None of my original code from the PoC is in Jamulus, AFAIK.

I also plan soon to update Opus from 1.4 to 1.5.2, as I saw mention that it had improved the packet-loss concealment. Of course, reducing packet loss due to jitter would also help a lot. In HPS's post linked above, he says "the biggest enemy for low latency audio is Jitter and not packet loss". In fact, packet sequencing compensates for jitter, and excess jitter then results in _effective_ packet loss due to packets not being physically lost, but being delayed enough to miss their slot completely and having to be discarded.

**Reply by dcorson-ticino-com** [reply]
> I don't know if a higher sampling rate or less compression is really that valuable? Is this really a major improvement?

I don't think that a higher sampling rate makes a difference for our application, but going to no compression makes a big difference.  Just try out the difference on sonobus, for example.  No compression is a audible big step up in the sound quality.

**Reply by dcorson-ticino-com** [reply]
> I think that we should rather focus on improving audio quality issues due to jitter. This is hard to achieve.

I don't know how many people have such poor connections these days to need extra measures to reduce the jitter.  Even in my transatlantic group, we rarely have connectivity problems and I don't see jitter as a problem.

**Reply by yeahletters** [reply]
@pljones #offtopic but @hselasky is sadly no longer with us. We will have to make do without his live expertise on this.

**Reply by pljones** [reply]
> @pljones #offtopic but @hselasky is sadly no longer with us. We will have to make do without his live expertise on this.

Ah, thanks 😞.

**Reply by iamthechrisb** [reply]
I only notice jitter issues with poor connections: this is usually a bad home network environment experiencing bufferbloat, really poor ISP service, or users trying to connect on Wifi or even mobile... With a healthy connection, and a good setup, this does not appear to be a problem. I'm feeling like the multi-port approach may not be warranted. Seems complex.

Audio quality is something I hear a great bit about from users. While the bandwidth profile for pcm may be a bit much for every client and server out there - the internet has changed a lot over the past few years -- and it will sound great. I hope to see this moved forward. Perhaps this is a good opportunity to make the code more modular with regard to codecs?

Opus is and was a great choice for Jamulus, I feel like that is the right choice for good sound on general internet connections. It would be great to see the updated version and have access to its capabilities. Recall https://github.com/orgs/jamulussoftware/discussions/1313 -- did this bring awareness to anything that needs revisited? Would it be possible to enable choosing between the normal encoder vs the custom encoder? Should the multichannel encoder be considered?

**Reply by ann0see** [reply]
> Perhaps this is a good opportunity to make the code more modular with regard to codecs?

Potentially yes.

>  It would be great to see the updated version and have access to its capabilities.

Yes. But the last time I tested upgrading it, it crashed Jamulus. This shouldn't happen and totally needs investigation.

> Should the multichannel encoder be considered?

Probably. However this **could** require a major redesign of Server and Client.

**Reply by softins** [reply]
(talking about Opus)

> > It would be great to see the updated version and have access to its capabilities.
> 
> Yes. But the last time I tested upgrading it, it crashed Jamulus. This shouldn't happen and totally needs investigation.

I have compiled with Opus 1.5.2 and it does indeed crash. I've tracked it down to a change introduced between 1.4 and 1.5, in `celt_encode_with_ec()` using a pointer that could be `NULL` without testing for that first. I've registered with https://gitlab.xiph.org to report the bug. Just waiting for my account to be approved.

**Reply by softins** [reply]
https://gitlab.xiph.org/xiph/opus/-/issues/2381

**Reply by ann0see** [reply]
Since the fix is available, we could apply it and test for any changes in our repo.

**Reply by ann0see** [reply]
We'd need to enable the ML features manually: https://opus-codec.org/demo/opus-1.5/

**Reply by softins** [reply]
> Since the fix is available, we could apply it and test for any changes in our repo.

I have a branch in which I am testing. Will raise a PR next week.

> We'd need to enable the ML features manually: https://opus-codec.org/demo/opus-1.5/

Haven't looked at what this is. Can be dealt with separately once 1.5.2+ is integrated.

**Reply by ann0see** [reply]
RAW audio was achieved.

**Comment by ann0see** [comment]
Thread for multi channel support

**Reply by ann0see** [reply]
Personally, I don't think we can bring multi channel support easily to Jamulus 4 due to the need of larger restructuring of core functionality of the code. I would rather not add this as to do.

**Reply by ann0see** [reply]
Can we decide to **not** have this initially as proposed Jamulus 4 feature?

**Reply by ann0see** [reply]
Since there hasn't been much discussion in this, I declare this not the main Jamulus 4 feature.

**Comment by ann0see** [comment]
Thread for access control

**Reply by ann0see** [reply]
Very controversial. Especially for public servers.

**Reply by rdica** [reply]
Hi all, yes it's controversial, and it still merits discussion because the idea of access control is not going away. :smiley: 

With the introduction of the RPC-JSON interface there are more possibilities for incorporating
client and server management.

In particular, access control.

While the majority of players and others are respectful of each other and the servers they
use, there are always bad actors, or unwanted/unauthorized clients/bots that
users/admins may want to manage. Sometimes muting isn't enough as that doesn't prevent the
bad actor from typing abuse in chat. There could be other reasons the group/admin may want
to manage connections to the server. It's also possible the admin may not
have access to manage firewalls and since the RPC-JSON interface is in userspace it doesn't
require root/admin access to manage it. One shouldn't have to be a firewall expert in the server
OS in order to manage client connections to their Jamulus servers.

@Rob-NY  has submitted PRs that add firewall functionality to the RPC-JSON interface, and I'm
advocating their inclusion to the code base.

I have been testing it on all my own servers for ~2 years and it works quite well, no
issues experienced to date.

One example; it could be incorporated into the server GUI, the admin could select a
client connection in the log dialog and click a button to manage it.

Other examples could utilize command line or web to perform the same operations.

Down the road, should the RPC-JSON interface have the ability to run from other than
localhost interface, specified clients could be given the RPC-JSON key and access
control (or any other supported management function provided by RPC-JSON) could be managed
directly from the authorized jamulus client itself.

I welcome constructive discussion regarding this topic, knowing how controversial it is.
Thanks for your consideration.

**Reply by yeahletters** [reply]
I would like to have at the very least access to a log of the chat's placed in the chat during jam sessions in a newer version. I have had on a one too many occasion well meaning or not individuals typing whatever they'd like into the chat. It is technically possible for one to type all kinds of profanities and abuse in the chat. I have since resorted to not reading chats anymore from certain individuals. I no longer feel safe, when random or anonymous people enter my own server when I am making music on my own, or jamming there. Can't for the life of me think why such a feature (moderated chat, or blacklisting certain IP addresses) would not be considered in a newer version.

**Reply by pljones** [reply]
Chat could be removed entirely without loss of core functionality, really.  I don't really like the lossy nature of Jamulus chat - it's "ok" but it's not exactly something that would attract me to the software or that I'd miss if it wasn't there.

**Reply by ann0see** [reply]
Saving chats has privacy implications and therefore was rejected. (Any client could save data written in the chat without anything noticing. And yes, sometimes phone numbers get exchanged). 

I would suggest to allow the server owner to be able to restrict the chat. Implementing this is not hard.

**Reply by pljones** [reply]
I'd still prefer to trim non-core features.  Using Matrix for chat seems like a better alternative and well aligned to how Jamulus works for audio.

**Reply by ann0see** [reply]
I would rather not remove the chat. It's a feature people use to e.g communicate in a non disturbing way. Since it's integrated it's convenient.

**Reply by dcorson-ticino-com** [reply]
I would not want the chat removed.  
Adding another program for chat functionality that most people will not be using and therefore not have installed sounds to just create problems.
The groups I am in do not use the chat often, but at times it has been a great help, also because everyone already has it up and running.

As to yeahletters problem above I'm wondering if muting someone also muting their chat entries would be a usable solution, including soloing etc.

**Reply by ann0see** [reply]
> if muting someone also muting their chat entries would be a usable solution, including soloing etc.

Interesting idea. 
I suggest something else however: a hide functionality.
Every user can hide a channel for themselves. If they hide it, the fader gets minimized somehow and any sound/chat message from this user gets blocked.

**Reply by pljones** [reply]
That's extra bloat in both client and server.  The client only knows messages are coming from the server, so the server has to know all these client preferences.  The client then needs to keep track of what it's asked the server to do so the UI can reflect this.

**Reply by softins** [reply]
The other important use for chat within a Jamulus session is for participants who do not have a mic, or even listeners who don't have any audio input. When I started with Jamulus, for a week or two before getting a proper interface, I just had my piano output direct into the computer, with no mic or mixer. I used chat to communicate with the others in the session.

Even recently while testing a new version, I was connected to a session with no audio input at all (laptop with no headphones, and the mic muted to avoid feedback), and used chat to introduce myself and make a couple of complimentary comments.

**Reply by rdica** [reply]
Passing URLs for charts/lyrics is a very common use.

**Reply by pljones** [reply]
Yes, I'm not saying "people shouldn't chat whilst using Jamulus".  Of course it's a useful (essential) thing to be able to do.  I'm saying it doesn't need to be part of Jamulus.  There are many other ways of chatting to people.

For example, one of the issues raised is being able to exclude chat from certain people.  One way would be for someone to start a channel on a discord or other chat server and, as they had admin privs, they'd be able to control who stayed in the chat.  None of that is to do with doing what Jamulus is meant for, hence I don't really want it in Jamulus.

**Reply by softins** [reply]
> Yes, I'm not saying "people shouldn't chat whilst using Jamulus". Of course it's a useful (essential) thing to be able to do. I'm saying it doesn't need to be part of Jamulus. There are many other ways of chatting to people.

I think many people (myself included) disagree strongly with this.

The whole point of the built-in chat is that _it's there_ automatically if you have joined a session. There is no need to find out what _other_ chat system the people in this particular session are using (there are so many to choose from, for most of which you need to register for an account).

It is also used to deliver the welcome message, which for many servers contains useful information such as contact details for the operator, purpose of the server, URL for a Jitsi session, etc.

**Reply by pljones** [reply]
> It is also used to deliver the welcome message, which for many servers contains useful information such as contact details for the operator, purpose of the server, URL for a Jitsi session, etc.

It could just do this - and nothing else.  And provide details of where to go for the chat, as well as video, conversations.

**Reply by ann0see** [reply]
In theory, yes. I think we agree that:
1. Jamulus Chat is not an absolute core feature
2. The chat is still necessary as it helps the community
3. Other chat applications may be superior and should be linked in the welcome message

Therefore, I vote to **not** remove the chat. However, I think we *can* include a JSON-RPC call to disable the chat during run time. I would prefer disabling the chat to be GUI/JSON-RPC, but not too easy to enable so that only people with experience can disable it.

**Reply by softins** [reply]
> > It is also used to deliver the welcome message, which for many servers contains useful information such as contact details for the operator, purpose of the server, URL for a Jitsi session, etc.
> 
> 
> 
> It could just do this - and nothing else.  And provide details of where to go for the chat, as well as video, conversations.

It could. I still think it's a bad idea. I also believe such a change would be incredibly unpopular.

**Reply by softins** [reply]
> In theory, yes. I think we agree that:
> 
> 1. Jamulus Chat is not an absolute core feature

In some situations it is pretty close to being a core feature.

> 2. The chat is still necessary as it helps the community

Yes.

> 3. Other chat applications may be superior and should be linked in the welcome message

They may be superior for some kinds of task, but not for on-the-fly messages within a session. And as I said above, there are many different chat applications, and different people use different ones, e.g. Discord, Signal, Telegram, WhatsApp, FB messenger. It would fragment communication, if some in the session only use one and others only use another. And maybe a few people don't subscribe to any.

> Therefore, I vote to **not** remove the chat.

Well I certainly agree with that.

> However, I think we *can* include a JSON-RPC call to disable the chat during run time. I would prefer disabling the chat to be GUI/JSON-RPC, but not too easy to enable so that only people with experience can disable it.

We could, I suppose. I expect the feature would get little use in the community at large.

**Reply by ignotus666** [reply]
Using an external medium for chatting still doesn't address the problem of morons posting offensive or disruptive messages - it just moves it somewhere else.
Another thing to consider is that there are many permanent servers where the server admin is hardly ever present. If blocking someone requires a server admin to do it, that's not much use if they're not present. I like the idea where muting a person's channel also mutes their chat messages. This allows anyone to choose whether they want to receive someone else's audio/chat without having to actually block anyone. Speaking as someone who hasn't a clue about how this works, but what about the client still receiving a muted person's chat messages, but replacing all the characters with '@' or similar?

**Reply by pljones** [reply]
Remember, the client doesn't do channel muting - it only knows it's asked the server to mute a channel.  Unless the server software supports a particular function ("chat muting"), then regardless of the client request, no change in server behaviour will be experienced.  And I'd expect that where server operators aren't making their presence available on their servers (e.g. contact details in the welcome message), they're also not updating their installation, so you'll not get any new server-side features.  There are a _lot_ of servers around that were installed at the start of Covid and have never been updated since, remember.

**Reply by ignotus666** [reply]
Yes, there are many servers running outdated versions. I was just thinking that maybe if the client has requested a muted channel, it could then render any chat messages received from that channel as a string of random characters - this wouldn't require any action from the server, would it?

**Reply by ann0see** [reply]
There's no need to send random characters. The server could just not send the message. 
Since this is for Jamulus 4, we can introduce features which do not 100% work on all versions before Jamulus 4.
I would say we just need to accept that this only works with updated servers.

**Reply by pljones** [reply]
There's been a reasonable uptake of 3.10.0 and 3.11.0 server versions, too.  Perhaps we could update the client to display the server version so the user had some idea of its capabilities and make their choice about which servers to use in a more informed manner, too.  Otherwise we could end up with "why does feature X not work for me but works for them?" questions with no clear way of answering that enables the user to make their own decisions.

**Reply by ann0see** [reply]
Possibly yes. We could put it in "About Jamulus" or settings. It shouldn't be on the mixer board.

**Reply by pljones** [reply]
I was thinking of the Connect dialog -- that's where people are going to make their choice about which server to connect to.  Being able to see whether or not it's the "kind of server" they want to join enables them to make the right decision.

**Reply by ann0see** [reply]
Yes. But not sure if we want to display the version there. Maybe on hover.

**Reply by dcorson-ticino-com** [reply]
I'm not sure why this must be a server feature,
The client knows the mute status of all the connected users it is sending to the server, why couldn't it just drop the chat messages it gets from the server of muted clients.
I agree that is not a clean solution, but would work for all server versions,

**Reply by ann0see** [reply]
The client does not know from which client a message came.

**Reply by softins** [reply]
> There's been a reasonable uptake of 3.10.0 and 3.11.0 server versions, too. Perhaps we could update the client to display the server version so the user had some idea of its capabilities and make their choice about which servers to use in a more informed manner, too. Otherwise we could end up with "why does feature X not work for me but works for them?" questions with no clear way of answering that enables the user to make their own decisions.

I proposed this and raised a PR over 3 years ago #1940, but it received too much friction and I gave up pushing it as it was too much like hard work. I still have the branch.

**Reply by ann0see** [reply]
The question is: does the user really care about the server version? Why isn't it enough to just not have the feature available/grayed out?

**Reply by softins** [reply]
> The question is: does the user really care about the server version? Why isn't it enough to just not have the feature available/grayed out?

It's information that might help them diagnose an issue, or that they can give to someone trying to help them, who may be more knowledgeable. I think it's useful, but wasn't willing to keep spending a lot of energy arguing for it.

**Reply by ann0see** [reply]
Ok. If you see it this way, they could also just check on explorer.

**Reply by softins** [reply]
But the point now is that my original feature was raised as a new suggestion further up this thread. That prompted me to mention my original PR, which did exactly what was suggested.

**Reply by pljones** [reply]
I'd rather see it in the connect dialog.  Like I say, that's where I'm looking when I make the decision as to what server to connect to.  It's got as much influence as the server capacity, I'd say, if it's going to determine what feature are available.

**Reply by softins** [reply]
Maybe, but that's a bit more involved. The client connect dialog would need also to send a `CLM_REQ_VERSION_AND_OS` to every listed server and process the `CLM_VERSION_AND_OS` responses, which it doesn't do currently.

**Reply by pljones** [reply]
Yes or the response to a request for the list of participants could be both the participant list and the `CLM_VERSION_AND_OS`, which would be ignored if unexpected and processed if expected.

**Reply by softins** [reply]
> Yes or the response to a request for the list of participants could be both the participant list and the `CLM_VERSION_AND_OS`, which would be ignored if unexpected and processed if expected.

The request for a participant list is only sent to a server that gave a non-zero client count in `CLM_PING_MS_WITHNUMCLIENTS` and a different count from the previous time. So with that scheme, empty servers would never send the version number. So it would be best if the connect dialog did explicitly request the version when sending a ping to a server, if the version of that server is not already known.

**Reply by pljones** [reply]
Mmm, okay so old servers are the limiting factor in compatibility for everything -- yeah, sounds like this is the only way.

**Reply by softins** [reply]
It should be doable ok on the client side. It's only the new version that wants to display the version that needs to send the request for it. It's just more complex than simply displaying the version of a connected server.

**Reply by pljones** [reply]
So we did server version.

The rest of the "access control" stuff seems to have faded away without a proposed, workable solution so far.

**Reply by mcfnord** [reply]
I hold strongly that a centralized defense is needed for the long-running, largely un-monitored servers that host most of the user-minutes listed on the seven built-in directories. But [my solution](https://github.com/mcfnord/jamulus/blob/jamfan/src/centraldefense.cpp) is considered unworkable by decision-makers, who don't want to maintain a central defense. It makes some sense to separate the vanilla technology from the more complicated and political niche case of public directory traffic anti-patterns. I tracked another griefer campaign last night, and felt proud that my dozen servers are immune. If attacks by bots or griefers increase, my server binary is a drop-in replacement for operators who want a hands-off solution. But perhaps we dally under a magical cloud of protection that keeps these anti-social patterns rare!

**Reply by ann0see** [reply]
It's likely that we could provide some simple management functionality over json rpc and this gets accepted

**Reply by ann0see** [reply]
I would propose to redesign the chat handling. The client should get channel id and text message and then parse locally. 

This would also enable hiding messages locally for new servers. Moreover it can fix various display bugs like invisible names in dark mode. 

We'd introduce a new enhanced chat message protocol message which gets sent to compatible clients.

**Reply by pljones** [reply]
I assume you mean the _source_ channel from which the chat message came.  Server generated messages and JSONRPC generated messages would need separate handling - but it could be as simple as assigning "special" channel ids.

**Reply by ann0see** [reply]
Yes. Sure the channel id the message originated from.

**Reply by ann0see** [reply]
Old:

```
- PROTMESSID_CHAT_TEXT: Chat text

    +------------------+----------------------+
    | 2 bytes number n | n bytes UTF-8 string |
    +------------------+----------------------+

    - "UTF-8 string": the chat message (plain text from client to server;
      HTML from server to client)
```

new (suggestion):
```
- PROTMESSID_CHAT_TEXT_CHANNEL: Chat text from channel
    | 1 byte channel ID | 2 bytes number n | n bytes UTF-8 string

```

Alternatively:

```
- PROTMESSID_CHAT_TEXT_CHANNEL: Chat text from channel
    | 1 byte channel ID | 64 bit (uint64_t) Unix timestamp | 2 bytes number n | n bytes UTF-8 string
```

**Reply by dingodoppelt** [reply]
A new protocol message is certainly the cleanest way.  
A dirty hack could just parse by the added html  
```
const QString strActualMessageText = "<font color=\"" + sCurColor + "\">(" + QTime::currentTime().toString ( "hh:mm:ss AP" ) + ") <b>" +
                                         ChanName.toHtmlEscaped() + "</b></font> " + strChatText.toHtmlEscaped();
```
and replace it with a different color.

**Reply by ann0see** [reply]
Maybe for legacy servers. But in the long term I think this new protocol message would be way cleaner and might in future allow for more features

**Reply by dingodoppelt** [reply]
> Maybe for legacy servers. But in the long term I think this new protocol message would be way cleaner and might in future allow for more features

Without a doubt and this should definitely happen.

**Reply by freenopen** [reply]
@rdica 
#### Clarification please
> the idea of access control is not going away.
>
When its only the same few people who request easy IP-block features then that does not tell anything about the real demand of the community. So from what do you conclude a general and priorized need of such IP-block features?

> There could be other reasons the group/admin may want to manage connections to the server. 
>
What *specific* and alternativeless reasons are that really?
And does the block of some bots really justify that deep change of the open community and giving some admins so much power over so many users?

#### Why that euphemism?
_I find the term "access control" as a camouflage euphemism anyway, because its just about simple IP-blocking, and that was mainly requested by a few persons to get rid of unwanted users. And by experience i can say that many of them who call for that feature here do *not* use it 'moderate' and with responsibility as written but just for excluding people who are either 'instrument competitors' or who have different opinions,  they only want to control who is allowed to play with who and who not. Means to gain personal advantage or to support their friends to gain 'e-fame', social rank or whatever - short: abuse._  

### An attempt to underlay some unchecked assumptions with data and evidence:
> While the majority of players and others are respectful of each other and the servers they use, there are always bad actors, or unwanted/unauthorized clients/bots that users/admins may want to manage. 
>
1. How many human '"bad actors" are that? (a number or atleast a percentage of users)
2. Is that number relevant atall in the waging of 'bad behaviour protection' versus implementing possible abuse features and giving up the 'open spirit' ?
3. How can a "bad actor" be identified; how is "bad behaviour" defined?
4. How relevant are human "bad actors" (whatever that is) for good working jams and software?
5. What other possibilities are there that do not offend base values and additionally open doors for systemic abuse?
6. Why is a mute not sufficient? I didn't hear any reasoning for that yet. (That someone could 'evade' to chat is not really a point because like for audio mute there is also no IP-block needed for chat mute.)
7.  To complete the already existing audio mute button - wouldn't it be easier and wouldn't it better comply to the open spirit and administrative abuse prevention to just implement such a button in the client mixer also for chat mute?
8. Why do some admins want "to manage" that and don't let each user decide for oneself, in particular when it can be realized by each user oneself? Are there other intentions than the pretended 'user protection'?
9. Why extend the power of one person over many others when its possible to let the individual decide?

### The alternative and easy solution
> Sometimes muting isn't enough as that doesn't prevent the bad actor from typing abuse in chat.
> 
But with that sentence you already give the solution!

**Add a chat mute button to the client mixer like the audio mute button!**

Thats my suggestion anyway already for years. So everybody can decide for oneself what he wants to read and what not.
We all are adult people, nobody needs an authority who decides for us with who we are allowed to play, to who we have to listen to or what we are allowed to read or not. 
IP-block for mute reasons is shooting on ants with canons. 
IP-block opens too many doors for systemic abuse, which we can already observe today on some servers and with some admins.
 
_( Edit: Personally i dont need a chat mute feature atall, because no words in a chat could harm me, its written words only, and i am open for everything, i can ignore it or talk to the one. But since some feel even harmed by someone using the word 'shit' in chat i prefer to serve their needs with a chat mute over introducing a huge systemic abuse feature and procedure into basic user rights and the spirit of open community)_

#### User bans on public services in europe have to comply EU laws
A public jamulus server is a public online service. You also shouldn't forget that according to GDPR and other laws, e.g. discrimination law, a public service provider or its representative (e.g. admin) has to give infos to atleast the banned user, which include:
1. **Block note** (info *that* the user is blocked.)
2. **Duration** (according to a given ruleset)
3. **Reason**. (reffering to a public ruleset that users have to confirm when entering the server, only no rulset obligatory when its an obvious law offense like e.g. racism, call for violance etc.)
4. **Prove**. (thats essential. e.g. screenshots or other documents. Just _"someone told me"_ is *no* prove *nor* a reason. .)

**No sentence/punishment/ban without reason and prove. That's _the_ fundamental principle in all european states' justice systems.**

If there is no prove or it can't be proven by principle then it's just slander/diffamation which is a criminal offense. An admin who would just follow such unproven claims by thirds would become involved as a co-conspirator.

_(Allone that the service needs to store information about/with the ban, already gives the affected user the right by GDPR to get all that info. Actually the GDPR does not only focus on 'stored' information on the system or online, but adresses **all** information the service provider has or keeps about the user. 
Additionally it is common and good habit to give info about the ban aswell to the other users of the service for several reasons.)_

If you implemement ban features you also have to ensure to implement features to comply to the above. Also all server admins would *have to* write and add a full ruleset or link to it to the server entry message and add a confirmation checkbox.
So why creating soooo much work when its not necessary atall because there are simple mute features?
Pure activism and the lust for coding (which i can understand very well :D ) shouldn't be relevant for a decision in such core issue.

Edit:
Above someone suggested a button where the user can make a user completely invisible int eh mixer including audio and chat mute in one - that would do it also. 
The technical question if the chat mute is done serverside or clientside is secondary imo, but i cannot judge that. 
Doing it clientside by just don't hand over the blocked users chattext to the chat window subroutine would probably be easy to implement, but would still keep traffic up when a user is spamming the chat with endless repetitive char garbage posts.

**Reply by pljones** [reply]
Note that "mute" is _not_ a Client operation: it's an instruction to the Server on how to behave when creating the mix you receive back.  I think it's technically possible to use the same or an additional setting on the Server to decide whether to send particular chat messages (i.e. the same idea: "don't include \<channel X\> in \<audio | chat\>").  Profile (shown in the mixer) is probably trickier.  (You'd need to keep the mixer channel there so you can unmute if you choose to.)

**Reply by dingodoppelt** [reply]
@freenopen
99% backing your statements here.
Years back I had to write [little tool](https://github.com/dingodoppelt/jamban) to IP block misbehaving users because exactly one person made that unavoidable (nothing but racist and derogatory comments). It works and there is definitely no need for an integration into Jamulus.
The potential for abuse grows immensely, once this gets available to everyone and therefore I'm against an integration into Jamulus.
As a part of the Jamulus community I know that 99.9% of the users are awesome. Jamban can easily handle the rest.
Speaking of "bad actors": Everybody knows one or two and it is always the same guys. We shouldn't give them too much publicity by wasting words and time and efforts on them because they are not in the position to break the Jamulus user experience (currently).
I have never experienced a session being destroyed by a bad actor and I don't know how frequently that happens, but I don't see an urgency to act now.
The one thing we do have control over is the official directory servers. I see more bad acting servers than I see misbehaving users so it might be wiser to think about mitigations on that level. I would leave single users out of the equation for now.

**Comment by ann0see** [comment]
Thread for plugin compatibility

**Reply by ann0see** [reply]
Probably doable. Requires a lot of implementation, I suppose. Feels like this could be done for Jamulus 4

**Reply by pljones** [reply]
I'm suggesting focussing on CLAP for the reason that it's an open spec without a restrictive licence agreement (Steinberg VST2 and 3 are both restrictive) and has cross-platform support (at least on Windows and macOS - don't know about Linux and probably not mobile).

**Reply by ann0see** [reply]
Seems as if this doesn't get that much vocal attention even though it's nice and probably doable. Any thoughts on concretizing CLAP support?
I'd like seeing this in Jamulus 4 though.

**Reply by pljones** [reply]
"Getting Started" page for developers: https://cleveraudio.org/developers-getting-started/ - very much oriented at plugin developers, though.
Host example (using rtaudio, which we'd rip out, of course): https://github.com/free-audio/clap-host?tab=readme-ov-file#minimal-clap-host

**Reply by ann0see** [reply]
@pljones Could you please raise an issue for CLAP support?
In it we should have:
1. A link to the CLAP repo
2. A clarification that client and server would have CLAP support
3. A general overview what needs to be done to make this happen

**Reply by pljones** [reply]
Done
https://github.com/jamulussoftware/jamulus/issues/3405

**Comment by iamthechrisb** [comment]
> In https://github.com/orgs/jamulussoftware/discussions/3341 we discussed potential larger features for Jamulus 4. This discussion serves as triaging the most important and actionable feature we can implement for Jamulus 4.
> 
> Step 1: Decide on primary feature to focus on Step 2: Plan how to get this feature going.
> 
> The best outcome would be to have an actionable plan as issue on what to work on exactly in form of a checklist.
> 
> We have 4 larger feature requests:
> 
>     1. Provide uncompressed/higher quality audio.
> 
>     2. Add multi channel support
> 
>     3. Implement Access control functionality
> 
>     4. Introduce plugin compatibility in Jamulus

These are all great ideas and the group should take ample time to weigh this out. Jamulus had great momentum during the pandemic, but development may need to shift gears to remain attractive to new users. Please don't take that negatively, Jamulus is a great work of software and the team here has done an amazing job. Everyone who has spent time playing is grateful. It's still amazing how well it works and how great it sounds for typical internet conditions.  

Has any thought been given to moving Jamulus to utilize the Juce audio framework? I had not seen this topic in the past, and through some basic searches.  Juce could allow the development effort make leaps in capabilities rather than building everything from the ground up while allowing core components, such as the Jamulus protocol, to get a fair balance of effort in refinement -- while bringing other capabilities to users. It is used by so many audio applications and very mature. Couldn't this help refocus efforts to get these features up and running -- are there license or cost issues with it?

**Reply by pljones** [reply]
Do you mean JUCE for plug-ins?  Or something else?

**Reply by iamthechrisb** [reply]
Not just for plug-ins, the whole project. This framework appears to be, at least able, to be the client.

**Reply by ann0see** [reply]
We're currently using Qt as framework and it seems a bit like https://juce.com/ also overlaps with some core functionality as far as I read the website. So it seems like a larger thing to do.

**Reply by pljones** [reply]
I guess if someone wanted to, they could use JUCE to write a client/server in using the published protocol.  But presumably, other than the message formats, it would be starting from scratch?

**Reply by dcorson-ticino-com** [reply]
I would see this as a potential base for a new MultiChannel-Jamulus.
My biggest worry would be the stability.  We presently own the protocol and know it is stable.  Using an external library we would depend on others.
I remember that Jamulus' nr1 characteristic is that it never crashes.

**Reply by iamthechrisb** [reply]
It's no small undertaking I'm sure, but I don't think this would be totally starting from scratch, hopefully it being c++ helps...  The important thing here is that its looked at and considered.  With discussions around a major release, I think its healthy for the project to look over this carefully. With its track record in audio, dsp capabilities, plugins - it could be a good fit. Further, Sonobus provides a solid study  that says it is stable for this type of application -- and has implemented some of the features being discussed.

**Reply by pljones** [reply]
If you can show an example of a Qt application using JUCE that running on Android, iOS, macOS, Linux and Windows (and supports JACK on both Linux and Windows - ideally we'd also support it on macOS), then it might be worth considering.  If we have to ditch Qt, then we're left with nothing but the protocol formats.  No code.

Then go check the amount of new code that was delivered in the last year and calculate how long adopting something new would take and decide if there might be a better use of that time elsewhere.

**Reply by dcorson-ticino-com** [reply]
> Sonobus provides a solid study that says it is stable for this type of application -- and has implemented some of the features being discussed.

Sonobus is the least stable of Jamulus' competition in my experience.  I would not use it as a good reference. Even JamKazam works better.

**Reply by ann0see** [reply]
I would rather prefer not to rewrite the whole app. Please note that moving to another framework should have been discussed in https://github.com/orgs/jamulussoftware/discussions/3341 instead.

See the initial post:

> This discussion serves as triaging the most important and actionable feature we can implement for Jamulus 4.

> Step 1: Decide on primary feature to focus on
> Step 2: Plan how to get this feature going.

**Reply by iamthechrisb** [reply]
> I would rather prefer not to rewrite the whole app. Please note that moving to another framework should have been discussed in https://github.com/orgs/jamulussoftware/discussions/3341 instead.
> 
> See the initial post:
> 
> > This discussion serves as triaging the most important and actionable feature we can implement for Jamulus 4.
> 
> > Step 1: Decide on primary feature to focus on
> > Step 2: Plan how to get this feature going.

that post probably needed more time

**Reply by ann0see** [reply]
The goal was to get to the most important and requested features for Jamulus 4. I expect that the most important ones would get in quite quickly.
We must triage and decide on what to do at some time. Having it around the release seemed sensible for me. You can always open a new discussion/issue.

**Comment by stefan1000** [comment]
Are there any plans to drop Qt 5 support in the 4.x release ? It seems to add unnecessary  complexity, and sticking with an outdated framework poses growing cybersecurity risks.

**Reply by pljones** [reply]
Qt6 adds UI breakage and platform incompatibility still, unfortunately.  It's being looked at in 4.x, though, yes.

**Reply by stefan1000** [reply]
Can you elaborate which platforms are incompatible with Qt6 or where the UI is broken?

**Reply by pljones** [reply]
Have a look through the PRs.  Or get Github CoPilot to do it.  That's what I'd have to do.

**Reply by ann0see** [reply]
iOS and macOS have issues on qt6

**Reply by softins** [reply]
I think we should keep Qt5 support for 4.0.0, but I would be happy to insist on 5.15+. I even have a branch with that restriction coded in `Jamulus.pro`. After #3797, all the supported versions of Debian and Ubuntu are using 5.15+.

**Reply by ann0see** [reply]
Yes. And iOS has a very hard time working correctly with Qt6
